File: src/main/java/net/ccbluex/liquidbounce/ui/client/clickgui/style/styles/fdpdropdown/FDPDropdownClickGUI.java
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge by LiquidBounce.
 * https://github.com/SkidderMC/FDPClient/
 */
package net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown;

import net.ccbluex.liquidbounce.features.module.Category;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.SideGui.SideGui;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.impl.SettingComponents;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.animations.Animation;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.animations.Direction;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.animations.impl.DecelerateAnimation;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.animations.impl.EaseBackIn;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.normal.Main;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.render.DrRenderUtils;
import net.ccbluex.liquidbounce.ui.client.hud.designer.GuiHudDesigner;
import net.ccbluex.liquidbounce.utils.render.RenderUtils;
import net.minecraft.client.gui.GuiScreen;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.util.ResourceLocation;
import org.lwjgl.input.Mouse;

import java.util.ArrayList;
import java.util.List;

public class FDPDropdownClickGUI extends GuiScreen {
    private final SideGui sideGui = new SideGui();
    public static String Name = "FDPClient";
    public static String version = "";
    public static String username;
    private Animation openingAnimation;
    private EaseBackIn fadeAnimation;
    private DecelerateAnimation configHover;
    private final ResourceLocation hudIcon = new ResourceLocation("fdpclient/custom_hud_icon.png");
    private List<MainScreen> categoryPanels;

    @Override
    public void initGui() {
        if (categoryPanels == null || Main.reloadModules) {
            categoryPanels = new ArrayList() {{
                for (Category category : Category.values()) {
                    //
                    add(new MainScreen(category));
                }
            }};
            Main.reloadModules = false;
        }
        sideGui.initGui();
        fadeAnimation = new EaseBackIn(400, 1, 2f);
        openingAnimation = new EaseBackIn(400, .4f, 2f);
        configHover = new DecelerateAnimation(250, 1);

        for (MainScreen catPanels : categoryPanels) {
            catPanels.animation = fadeAnimation;
            catPanels.openingAnimation = openingAnimation;
            catPanels.initGui();
        }
    }

    @Override
    protected void keyTyped(char typedChar, int keyCode) {
        if (keyCode == 1) {
            openingAnimation.setDirection(Direction.BACKWARDS);
        //    sideGui.focused = false;
            fadeAnimation.setDirection(openingAnimation.getDirection());
        }
        sideGui.keyTyped(typedChar, keyCode);
        categoryPanels.forEach(categoryPanel -> categoryPanel.keyTyped(typedChar, keyCode));
    }

    @Override
    public boolean doesGuiPauseGame() {
        return false;
    }

    @Override
    public void drawScreen(int mouseX, int mouseY, float partialTicks) {
        if (Mouse.isButtonDown(0) && mouseX >= 5 && mouseX <= 50 && mouseY <= height - 5 && mouseY >= height - 50)
            mc.displayGuiScreen(new GuiHudDesigner());
        RenderUtils.INSTANCE.drawImage(hudIcon, 9, height - 41, 32, 32);

        if (Main.reloadModules) {
            initGui();
        }

        if (openingAnimation.isDone() && openingAnimation.getDirection().equals(Direction.BACKWARDS)) {
            mc.displayGuiScreen(null);
            return;
        }

        boolean focusedConfigGui = sideGui.focused;
        int fakeMouseX = focusedConfigGui ? 0 : mouseX, fakeMouseY = focusedConfigGui ? 0 : mouseY;
        ScaledResolution sr = new ScaledResolution(mc);

        boolean hoveringConfig = DrRenderUtils.isHovering(width - 120, height - 65, 75, 25, fakeMouseX, fakeMouseY);

        configHover.setDirection(hoveringConfig ? Direction.FORWARDS : Direction.BACKWARDS);
        int alphaAnimation = Math.max(0, Math.min(255, (int) (255 * fadeAnimation.getOutput())));

        GlStateManager.color(1, 1, 1, 1);

        SettingComponents.scale = (float) (openingAnimation.getOutput() + .6f);
        DrRenderUtils.scale(sr.getScaledWidth() / 2f, sr.getScaledHeight() / 2f, (float) openingAnimation.getOutput() + .6f, () -> {
            for (MainScreen catPanels : categoryPanels) {
                catPanels.drawScreen(fakeMouseX, fakeMouseY);
            }

            sideGui.drawScreen(mouseX, mouseY, partialTicks, alphaAnimation);
        });


    }

    @Override
    protected void mouseClicked(int mouseX, int mouseY, int mouseButton) {
        boolean focused = sideGui.focused;
        sideGui.mouseClicked(mouseX, mouseY, mouseButton);
        if (!focused) {
            categoryPanels.forEach(cat -> cat.mouseClicked(mouseX, mouseY, mouseButton));
        }
    }

    @Override
    protected void mouseReleased(int mouseX, int mouseY, int state) {
        boolean focused = sideGui.focused;
        sideGui.mouseReleased(mouseX, mouseY, state);
        if (!focused) {
            categoryPanels.forEach(cat -> cat.mouseReleased(mouseX, mouseY, state));
        }
    }

}

File: src/main/java/net/ccbluex/liquidbounce/ui/client/clickgui/style/styles/fdpdropdown/FDPDropdownClickGUI.kt
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/SkidderMC/FDPClient/
 */
package net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown

import net.ccbluex.liquidbounce.FDPClient.CLIENT_NAME
import net.ccbluex.liquidbounce.features.module.Category
import net.ccbluex.liquidbounce.features.module.modules.client.ClickGUIModule
import net.ccbluex.liquidbounce.features.module.modules.client.HUDModule.guiColor
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.SideGui.SideGui
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.impl.SettingComponents
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.animations.Animation
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.animations.Direction
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.animations.impl.DecelerateAnimation
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.animations.impl.EaseBackIn
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.normal.Main
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.utils.render.DrRenderUtils
import net.ccbluex.liquidbounce.ui.client.hud.designer.GuiHudDesigner
import net.ccbluex.liquidbounce.ui.font.AWTFontRenderer.Companion.assumeNonVolatile
import net.ccbluex.liquidbounce.utils.render.RenderUtils
import net.ccbluex.liquidbounce.utils.render.RenderUtils.drawBloom
import net.minecraft.client.gui.GuiScreen
import net.minecraft.client.gui.ScaledResolution
import net.minecraft.util.ResourceLocation
import org.lwjgl.input.Mouse
import java.awt.Color

/**
 * ClickGUI FDP
 */
class FDPDropdownClickGUI : GuiScreen() {

    private val sideGui = SideGui()

    private lateinit var openingAnimation: Animation
    private lateinit var fadeAnimation: EaseBackIn
    private lateinit var configHover: DecelerateAnimation

    private val hudIcon = ResourceLocation("${CLIENT_NAME.lowercase()}/custom_hud_icon.png")

    private var categoryPanels: MutableList<DropdownCategory>? = null
    private var isInitialized = false

    override fun initGui() {
        try {
            if (categoryPanels == null || Main.reloadModules || !isInitialized) {
                categoryPanels = mutableListOf<DropdownCategory>().apply {
                    Category.entries.forEach { category ->
                        add(DropdownCategory(category))
                    }
                }
                Main.reloadModules = false
                isInitialized = true
            }

            sideGui.initGui()

            fadeAnimation = EaseBackIn(400, 1.0, 2.0f)
            openingAnimation = EaseBackIn(400, 0.4, 2.0f)
            configHover = DecelerateAnimation(250, 1.0)

            categoryPanels?.forEach { panel ->
                panel.animation = fadeAnimation
                panel.openingAnimation = openingAnimation
                panel.initGui()
            }
        } catch (e: Exception) {
            println("Error initializing FDPDropdownClickGUI: ${e.message}")
        }
    }

    override fun keyTyped(typedChar: Char, keyCode: Int) {
        try {
            if (keyCode == 1) {
                openingAnimation.direction = Direction.BACKWARDS
                fadeAnimation.direction = openingAnimation.direction
            }
            sideGui.keyTyped(typedChar, keyCode)
            categoryPanels?.forEach { it.keyTyped(typedChar, keyCode) }
        } catch (e: Exception) {
            println("Error handling key input: ${e.message}")
        }
    }

    override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float) {
        try {
            assumeNonVolatile {
                if (Mouse.isButtonDown(0) && mouseX in 5..50 && mouseY in (height - 50)..(height - 5)) {
                    mc.displayGuiScreen(GuiHudDesigner())
                }
                RenderUtils.drawImage(hudIcon, 9, height - 41, 32, 32)
                if (openingAnimation.isDone && openingAnimation.direction == Direction.BACKWARDS) {
                    mc.displayGuiScreen(null)
                    return@assumeNonVolatile
                }
                val sr = ScaledResolution(mc)
                val finalScale = (openingAnimation.output + 0.6f) * ClickGUIModule.scale
                SettingComponents.scale = finalScale.toFloat()
                val transformedMouseX = sr.scaledWidth / 2f + (mouseX - sr.scaledWidth / 2f) / finalScale
                val transformedMouseY = sr.scaledHeight / 2f + (mouseY - sr.scaledHeight / 2f) / finalScale

                val focusedConfigGui = sideGui.focused
                val effectiveMouseX = if (focusedConfigGui) 0 else transformedMouseX.toInt()
                val effectiveMouseY = if (focusedConfigGui) 0 else transformedMouseY.toInt()

                DrRenderUtils.scale(sr.scaledWidth / 2f, sr.scaledHeight / 2f, finalScale.toFloat()) {
                    categoryPanels?.forEach { it.drawScreen(effectiveMouseX, effectiveMouseY) }
                }
                sideGui.drawScreen(mouseX, mouseY, partialTicks, (255 * fadeAnimation.output).toInt().coerceIn(0, 255))
            }
            drawBloom(mouseX - 5, mouseY - 5, 10, 10, 16, Color(guiColor, true))
            super.drawScreen(mouseX, mouseY, partialTicks)
        } catch (e: Exception) {
            println("Error during FDP GUI rendering: ${e.message}")
        }
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        try {
            val oldFocus = sideGui.focused
            sideGui.mouseClicked(mouseX, mouseY, mouseButton)
            if (!oldFocus) {
                val sr = ScaledResolution(mc)
                val finalScale = (openingAnimation.output + 0.6f) * ClickGUIModule.scale
                val transformedMouseX = sr.scaledWidth / 2f + (mouseX - sr.scaledWidth / 2f) / finalScale
                val transformedMouseY = sr.scaledHeight / 2f + (mouseY - sr.scaledHeight / 2f) / finalScale
                categoryPanels?.forEach { it.mouseClicked(transformedMouseX.toInt(), transformedMouseY.toInt(), mouseButton) }
            }
        } catch (e: Exception) {
            println("Error handling mouse click: ${e.message}")
        }
    }

    override fun mouseReleased(mouseX: Int, mouseY: Int, state: Int) {
        try {
            val oldFocus = sideGui.focused
            sideGui.mouseReleased(mouseX, mouseY, state)
            if (!oldFocus) {
                val sr = ScaledResolution(mc)
                val finalScale = (openingAnimation.output + 0.6f) * ClickGUIModule.scale
                val transformedMouseX = sr.scaledWidth / 2f + (mouseX - sr.scaledWidth / 2f) / finalScale
                val transformedMouseY = sr.scaledHeight / 2f + (mouseY - sr.scaledHeight / 2f) / finalScale
                categoryPanels?.forEach { it.mouseReleased(transformedMouseX.toInt(), transformedMouseY.toInt(), state) }
            }
        } catch (e: Exception) {
            println("Error handling mouse release: ${e.message}")
        }
    }

    fun resetGui() {
        isInitialized = false
        categoryPanels?.clear()
        categoryPanels = null
    }

    override fun doesGuiPauseGame() = false
}